File: algorithms/Delete Node in a BST/Delete Node in a BST.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */

const getLeftMostNode=(root)=>{
    if(!root)return null;
    let node=getLeftMostNode(root.left);
    return node?node:root;
}
/**
 * @param {TreeNode} root
 * @param {number} key
 * @return {TreeNode}
 */


var deleteNode = function(root, key) {
    if(!root)return null;
    if(root.val>key){
    root.left=deleteNode(root.left,key);
    }else if(root.val<key){
    root.right= deleteNode(root.right,key);
    }else{
        if(!root.left)return root.right;
        if(!root.right)return root.left;
        let succ_node=getLeftMostNode(root.right);
        root.val=succ_node.val;
        root.right= deleteNode(root.right,succ_node.val);
        
    }
    return root;

};

File: scripts/algorithms/D/Delete Node in a BST/Delete Node in a BST.java
// Runtime: 0 ms (Top 100.0%) | Memory: 44.92 MB (Top 83.3%)

class Solution {
    public TreeNode deleteNode(TreeNode root, int key) {
        if(root==null) return null;
        
        if(key<root.val){                            
            root.left = deleteNode(root.left,key);
            return root;
        }
        
        else if(key>root.val){
            root.right = deleteNode(root.right,key);
            return root;
        }
        
        else{
            if(root.left==null){
                return root.right;
            }
            else if(root.right==null){
                return root.left;
            }
            else{
                TreeNode min = root.right;
                while(min.left!=null){
                    min = min.left;
                }
                
                root.val = min.val;
                root.right = deleteNode(root.right,min.val);
                return root;
            }
        }
    }
}